src/test/App.test.js

```diff
@@ -1,41 +1,41 @@
-import renderer from "react-test-renderer";
-import { render } from "@testing-library/react";
+import { render, screen } from "@testing-library/react";
 
 import App from "../App"
+jest.mock('../components/Header', () => {
+  return () => <div>HeaderMock</div>
+});
+jest.mock('../components/Splash', () => {
+  return () => <div>SplashMock</div>
+});
+jest.mock('../components/About', () => {
+  return () => <div>AboutMock</div>
+});
+jest.mock('../components/Events', () => {
+  return () => <div>EventsMock</div>
+});
+jest.mock('../components/Members', () => {
+  return () => <div>MembersMock</div>
+});
+jest.mock('../components/Footer', () => {
+  return () => <div>FooterMock</div>
+});
 
 describe("App", () => {
-  it("renders", () => {
-    const view = renderer.create(<App />);
-    expect(view).toMatchSnapshot();
-  });
-
-  describe("renders links to external sites", () => {
-    it.each([
-      ['freeCodeCamp.org homepage', 'freecodecamp.org'],
-      ['Discord homepage', 'discord.com/invite/EXehPVnBYz'],
-      ['FCCColumbus Github profile', 'github.com/FCCColumbus'],
-      ['FCCColumbus Github contributors', 'github.com/FCCColumbus/cbus-web/graphs/contributors'],
-    ])(`%s`, (_, expected) => {
-      const { container } = render(<App />);
-      const anchorElements = container.querySelectorAll('a');
+  test("renders App components in order", () => {
+    //first, render the app as we mocked it
+    render(<App />)
 
-      const actual = Object.values(anchorElements)
-        .map(anchorElement => anchorElement.attributes.href.value)
-        .some(value => value.includes(expected));
+    //make sure the Header component is there (since we use it below)
+    expect(screen.getByText('HeaderMock')).toBeInTheDocument(); 
 
-      expect(actual).toBeTruthy();
-    });
-  });
+    //the Header's parent element should be the App HTMLElement
+    const appnode = screen.getByText('HeaderMock').parentElement;
 
-  test("has an Events section", () => {
-    const { container } = render(<App />);
-    const event = container.querySelector('#events');
-    expect(event).toBeInTheDocument();
+    //at this point we can just do a string match on the raw outerHTML of the app element
+    //we do this by using a regular expression (RegEx).  The {1,1} notation means "one and only one"
+    //and the [\s\S]* ignores any strings in between
+    //note the order of the component mocks.  when the order is changed in App.js, this test will fail (as it should)
+    expect(appnode.outerHTML).toMatch(/HeaderMock{1,1}[\s\S]*SplashMock{1,1}[\s\S]*AboutMock{1,1}[\s\S]*EventsMock{1,1}[\s\S]*MembersMock{1,1}[\s\S]*FooterMock{1,1}/s);
```

Comment:
One issue with this assertion is that the intended effect of `{1,1}` is immediately cancelled by `[\s\S]*`. The assertion as written does verify that the Header, Splash, About, etc components exist in the desired order but doesn't fail if for example we render the Splash component any additional number of times anywhere after the Header and before the Footer -- i.e. anywhere that this additional rendering would be matched by `[\s\S]*` ("any character any amount of times, including newlines").

Taken pedantically, the assertion does in a sense match with the name of the test, but it seems that we ought to also have some way to more comprehensively test that "the App component renders exactly what we want it to" instead of "the App component renders these specific components in this order, and might also render some other stuff in there too".

I wrote a few commits on how I might approach this but got a little confused on where to push the branch so I'll mention it in the Discord.

Reply:
@jduffey looks good.  

can you merge your changes from your [test branch](https://github.com/jduffey/cbus-web/commits/jed-branch-test) to here?

Reply:
Done ✅ Didn't wanna push directly onto this branch without permission/discussion first so thanks for looking them over.